File: benchmarks/executable/restaurant_uint_saturating_policy.c
// C23 reference for the Restaurant UInt saturating-policy workload.

#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>

#ifdef _WIN32
#include <fcntl.h>
#include <io.h>
#endif

static volatile uint64_t runtime_negate_zero = UINT64_C(0);
static volatile uint64_t runtime_negate_maximum = UINT64_MAX;
static volatile uint64_t runtime_ordinary_base = UINT64_C(2);
static volatile uint64_t runtime_ordinary_exponent = UINT64_C(3);
static volatile uint64_t runtime_clamped_base = UINT64_C(2);
static volatile uint64_t runtime_clamped_exponent = UINT64_C(64);
static volatile uint64_t runtime_identity_base = UINT64_C(0);
static volatile uint64_t runtime_identity_exponent = UINT64_C(0);

static uint64_t saturating_negate_u64(uint64_t value) {
  const uint64_t wrapped = UINT64_C(0) - value;
  return wrapped > UINT64_C(0) ? UINT64_C(0) : wrapped;
}

static uint64_t saturating_multiply_u64(uint64_t left, uint64_t right) {
  if (left != UINT64_C(0) && right > UINT64_MAX / left) return UINT64_MAX;
  return left * right;
}

static uint64_t saturating_power_u64(uint64_t base, uint64_t exponent) {
  uint64_t result = UINT64_C(1);
  while (exponent != UINT64_C(0)) {
    if ((exponent & UINT64_C(1)) != UINT64_C(0))
      result = saturating_multiply_u64(result, base);
    exponent >>= 1;
    if (exponent != UINT64_C(0))
      base = saturating_multiply_u64(base, base);
  }
  return result;
}

int main(void) {
#ifdef _WIN32
  if (_setmode(_fileno(stdout), _O_BINARY) == -1) return 1;
#endif
  const uint64_t neg_zero = saturating_negate_u64(runtime_negate_zero);
  const uint64_t neg_maximum = saturating_negate_u64(runtime_negate_maximum);
  const uint64_t ordinary = saturating_power_u64(
      runtime_ordinary_base, runtime_ordinary_exponent);
  const uint64_t clamped = saturating_power_u64(
      runtime_clamped_base, runtime_clamped_exponent);
  const uint64_t identity = saturating_power_u64(
      runtime_identity_base, runtime_identity_exponent);
  return printf("Saturating policy %" PRIu64 "/%" PRIu64 "/%" PRIu64
                "/%" PRIu64 "/%" PRIu64 "\n",
                neg_zero, neg_maximum, ordinary, clamped, identity) < 0
      ? 1
      : 0;
}

File: benchmarks/executable/restaurant_uint_saturating_policy.rs
// Rust 2024 reference for the Restaurant UInt saturating-policy workload.
// Expected exit: 0
// Expected stdout:
// Saturating policy add 18446744073709551615/11; subtract 0/10; multiply 18446744073709551615/42; negate 0/0; power 8/18446744073709551615/1

use std::hint::black_box;
use std::io::{self, Write};

fn saturating_add_u64(left: u64, right: u64) -> u64 {
    left.saturating_add(right)
}

fn saturating_subtract_u64(value: u64, amount: u64) -> u64 {
    value.saturating_sub(amount)
}

fn saturating_negate_u64(value: u64) -> u64 {
    0_u64.saturating_sub(value)
}

fn saturating_multiply_u64(left: u64, right: u64) -> u64 {
    left.saturating_mul(right)
}

fn saturating_power_u64(mut base: u64, mut exponent: u64) -> u64 {
    let mut result = 1_u64;
    while exponent != 0 {
        if exponent & 1 != 0 {
            result = result.saturating_mul(base);
        }
        exponent >>= 1;
        if exponent != 0 {
            base = base.saturating_mul(base);
        }
    }
    result
}

fn main() {
    let maximum_add = saturating_add_u64(black_box(u64::MAX), black_box(1_u64));
    let ordinary_add = saturating_add_u64(black_box(10_u64), black_box(1_u64));
    let underflow_subtract =
        saturating_subtract_u64(black_box(0_u64), black_box(1_u64));
    let ordinary_subtract =
        saturating_subtract_u64(black_box(11_u64), black_box(1_u64));
    let overflow_multiply =
        saturating_multiply_u64(black_box(u64::MAX), black_box(2_u64));
    let ordinary_multiply =
        saturating_multiply_u64(black_box(6_u64), black_box(7_u64));
    let neg_zero = saturating_negate_u64(black_box(0_u64));
    let neg_maximum = saturating_negate_u64(black_box(u64::MAX));
    let ordinary = saturating_power_u64(black_box(2_u64), black_box(3_u64));
    let clamped = saturating_power_u64(black_box(2_u64), black_box(64_u64));
    let identity = saturating_power_u64(black_box(0_u64), black_box(0_u64));
    let mut stdout = io::stdout().lock();
    if writeln!(
        stdout,
        "Saturating policy add {maximum_add}/{ordinary_add}; subtract {underflow_subtract}/{ordinary_subtract}; multiply {overflow_multiply}/{ordinary_multiply}; negate {neg_zero}/{neg_maximum}; power {ordinary}/{clamped}/{identity}",
    )
    .is_err()
    {
        std::process::exit(1);
    }
}
